### PR TITLE
[CCTRI-2086] APOD API

### PR DESCRIPTION
**Pull request by mstoro**

### CODE REVIEW

**Review by AnastasiiaRozlyvan — COMMENTED** [review]
Hi! Please check your code with `flake8` tool and fix the warnings if any.

**Comment by mstoro** [comment]
> Hi! Please check your code with `flake8` tool and fix the warnings if any.

Hi! Fixed.

File: app/app.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from flask import Flask
+
+from api.apod import apod_api
+from auth import auth
+
+from db import db
+app = Flask(__name__)
+
+app.config['SECRET_KEY'] = 'asd[jdsvcx3'
```
Comment (AnastasiiaRozlyvan) [inline app/app.py]: It seems to me that it is not very safe to store the secret key in app.py file. Perhaps you should move it to environment variables?

I also have a suggestion to put all app.config values into a separate Config file

Comment (mstoro) [reply app/app.py]: https://github.com/mstoro/nasa_api/pull/1/commits/e7a312b2874fcbc81eb477d9ba5760e56419ee07

File: reqirements.txt [hunk 2]
```diff
@@ -0,0 +1,41 @@
+alembic==1.5.4
```
Comment (AnastasiiaRozlyvan) [inline reqirements.txt]: You can try to shorten this list. I think that some dependencies from this list is extra and will be installed automatically with top-level dependencies.

File: wrapers.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import jwt
+from flask import request, jsonify, current_app
+from functools import wraps
+
+
+from db import User
+
+
+def token_required(f):
```
Comment (AnastasiiaRozlyvan) [inline wrapers.py]: To make it more readable, it might be worth replacing `f` with `func`, for example.

File: reqirements.txt (hunk 2 above)
Comment (mstoro) [reply reqirements.txt]: https://github.com/mstoro/nasa_api/pull/1/commits/dd4ea7f64b13c3b441a41ee185d61f81b7d20393

File: wrapers.py (hunk 3 above)
Comment (mstoro) [reply wrapers.py]: https://github.com/mstoro/nasa_api/pull/1/commits/606c73d5d0205ad63bdbcc79471c67f249994c86

**Comment by AnastasiiaRozlyvan** [comment]
And the last thing that would be nice to do here is to create a file structure like [this](https://github.com/CiscoSecurity/tr-05-serverless-relay). Because the goal of training is just to learn how to make such applications. I understand that not all of this can be applied to APOD API, but I think you can at least get rid of app folder, rename main.py -> app.py etc.  

Also note that we have the `develop` branch in the repositories on our project and we merge our commits into this branch. We only merge releases into the master branch.  Let it stay as it is in this repository, but when you do real tasks do not forget about it)

**Comment by mstoro** [comment]
> And the last thing that would be nice to do here is to create a file structure like [this](https://github.com/CiscoSecurity/tr-05-serverless-relay). Because the goal of training is just to learn how to make such applications. I understand that not all of this can be applied to APOD API, but I think you can at least get rid of app folder, rename main.py -> app.py etc.
> 
> Also note that we have the `develop` branch in the repositories on our project and we merge our commits into this branch. We only merge releases into the master branch. Let it stay as it is in this repository, but when you do real tasks do not forget about it)

https://github.com/mstoro/nasa_api/pull/1/commits/7d4991cc6800ecbe2a30b4fbe13307e171d41996

**Review by AnastasiiaRozlyvan — APPROVED** [review]

File: api/apod.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+from api.request import Request
+from api.wrappers import token_required
+
+apod_api = Blueprint('apod', __name__)
+
+
+@apod_api.route('/apod', methods=['POST'])
+@token_required
+def apod(current_user):
+    url = 'https://api.nasa.gov/planetary/apod'
```
Comment (AlinaBlahun) [inline api/apod.py]: Looks like something worth moving to config

**Review by AlinaBlahun — COMMENTED** [review]
According to PEP8
Imports should be grouped in the following order:
- Standard library imports.
- Related third party imports.
- Local application/library specific imports.
You should put a blank line between each group of imports. 
https://www.python.org/dev/peps/pep-0008/#imports.
You can use PyCharm tools to automate that https://www.jetbrains.com/help/pycharm/creating-and-optimizing-imports.html#optimize-imports-in-project e.g.  `Ctrl+Alt+O` shortcut.

File: api/apod.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+from api.schemas import PayloadSchema
+from api.request import Request
+from api.wrappers import token_required
+
+apod_api = Blueprint('apod', __name__)
+
+
+@apod_api.route('/apod', methods=['POST'])
+@token_required
+def apod(current_user):
```
Comment (AlinaBlahun) [inline api/apod.py]: Is current_user used somewere?

File: api/utils.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import requests
+from PIL import Image
+from io import BytesIO
+
+
+def make_request(url, params=None):
+    response = requests.get(url, params=params)
+    return response.json()
```
Comment (AlinaBlahun) [inline api/utils.py]: Is it used?

File: api/utils.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+
+def get_params(request):
+    params = {'api_key': 'DEMO_KEY',
```
Comment (AlinaBlahun) [inline api/utils.py]: Let's move `api_key` to environment variables

File: api/utils.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+from PIL import Image
+from io import BytesIO
+
+
+def make_request(url, params=None):
+    response = requests.get(url, params=params)
+    return response.json()
+
+
+def get_params(request):
```
Comment (AlinaBlahun) [inline api/utils.py]: Are you sure we need this function? Try to do the same in PayloadSchema with marshmallow tools (default or missing params) If there is no need in api_key validation as it's constant you can just add it to PayloadSchema validation result

File: api/apod.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+    apod_request = Request(url, params=params)
+
+    response = apod_request.get().json()
+
+    if 'thumbnail_url' in response:
+        image_url = response['thumbnail_url']
+    else:
+        image_url = response['hdurl'] if bool(params['hd']) is True \
+            else response['url']
```
Comment (AlinaBlahun) [inline api/apod.py]: Avoid backslashes, please.
The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation. https://www.python.org/dev/peps/pep-0008/#maximum-line-length

File: api/utils.py (hunk 6 above)
Comment (mstoro) [reply api/utils.py]: https://github.com/mstoro/nasa_api/pull/1/commits/aa95766f824ad79eea1e616ef7c7ca1b03b6f8ea

File: api/wrappers.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+        try:
+            data = jwt.decode(
+                token,
+                current_app.config['SECRET_KEY'],
+                algorithms=['HS256']
+            )
+            current_user = User.query.filter_by(id=data['id']).first()
+        except jwt.ExpiredSignatureError:
+            return jsonify({'message': 'Your token expired! \
```
Comment (AlinaBlahun) [inline api/wrappers.py]: As you can see, `jsonify({'message': ..` is everywhere, let's do something with it. At least move that to some function e.g jsonify_message in utils like [here](https://github.com/CiscoSecurity/tr-05-serverless-google-safe-browsing/blob/b656485707089190e1c805b92ac22b60b57659ea/code/api/utils.py#L138)

Also, you can create custom exception with the same logic and do some app level handler for it. Please take a look [here](https://github.com/CiscoSecurity/tr-05-serverless-farsight-dnsdb/blob/04b14a46bc05d108ab6f43a43b3e774c563aeb98/api/errors.py ) and [here](https://github.com/CiscoSecurity/tr-05-serverless-farsight-dnsdb/blob/04b14a46bc05d108ab6f43a43b3e774c563aeb98/app.py#L36)

File: api/wrappers.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        if not token:
+            return jsonify({'message': 'Token is missing!'})
+
+        try:
+            data = jwt.decode(
+                token,
+                current_app.config['SECRET_KEY'],
+                algorithms=['HS256']
+            )
+            current_user = User.query.filter_by(id=data['id']).first()
```
Comment (AlinaBlahun) [inline api/wrappers.py]: I would create function in db for actions like that

File: api/wrappers.py (hunk 10 above)
Comment (mstoro) [reply api/wrappers.py]: https://github.com/mstoro/nasa_api/pull/1/commits/0217f91d71366d0c783fcf70ff61e1512d8ef806

File: api/schemas.py [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+from marshmallow import Schema, fields, ValidationError
+
+
+def validate_date(given_date):
+    if given_date > date.today():
+        raise ValidationError('Enter existing date')
+
+
+class PayloadSchema(Schema):
+    api_key = fields.String(missing='DEMO_KEY')
```
Comment (AlinaBlahun) [inline api/schemas.py]: Move it 'DEMO_KEY' to config, please

File: app.py [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+app.register_blueprint(apod_api)
+
+db.init_app(app)
+
+
+@app.errorhandler(TokenError)
+def handle_token_errors(error):
+    app.logger.error(error.json)
+    g.errors = [error.json]
+    return jsonify(g.errors)
```
Comment (AlinaBlahun) [inline app.py]: You can do simply `return jsonify(error.json)`. We used `g` in the example as an accumulator for results, you don't need it here.

File: api/schemas.py (hunk 12 above)
Comment (mstoro) [reply api/schemas.py]: https://github.com/mstoro/nasa_api/pull/1/commits/0b16b705764a94293e87719fd037611c8fcab3c5

**Review by AlinaBlahun — CHANGES_REQUESTED** [review]
Sorry, I just read your task and there is no a word about user sign-up, and also image should be saved to the folder, and the link for that folder returned. @mstoro it looks like you didn't understand the task, so there a lot of things to redo!

I think it’s a great lesson for all of us that any development should start with requirements analysis, cause errors done on this step are the most expensive ones!

I am sure that Maksym Honchar is the best candidate to guide you through the task idea as he has done this task before. Please have a short meeting to analyze what should be done here. Thank you, guys!

File: api/errors.py [hunk 14]
```diff
@@ -1,13 +1,29 @@
...
+    def __init__(self):
+        super().__init__('Invalid api key', HTTPStatus.BAD_REQUEST)
+
+
+class RequestError(ApiKeyError):
+    def __init__(self, message):
+        super().__init__(message, status=HTTPStatus.BAD_REQUEST)
+
+
+# class MissingApiKeyError(ApiKeyError):
```
Comment (AnastasiiaRozlyvan) [inline api/errors.py]: I think it is better to delete these comments

Comment (mstoro) [reply api/errors.py]: https://github.com/mstoro/nasa_api/pull/1/commits/72d016018bdc3cc1c5284981c061931570064aea

File: schema.sql [hunk 15]
```diff
@@ -0,0 +1,7 @@
+DROP TABLE IF EXISTS user;
+
+CREATE TABLE user(
+    id INTEGER PRIMARY KEY AUTOINCREMENT,
+    name TEXT UNIQUE NOT NULL,
+    password TEXT NOT NULL
+);
```
Comment (AlinaBlahun) [inline schema.sql]: Are you sure you still need this?

Comment (mstoro) [reply schema.sql]: No, deleted it. https://github.com/mstoro/nasa_api/pull/1/commits/fc29d10da877dcf14722e0f9111d975be3acde1b

File: api/apod.py [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+    apod_request = Request(url, params=params)
+
+    response = apod_request.get().json()
+
+    if 'thumbnail_url' in response:
+        image_url = response['thumbnail_url']
+    else:
+        image_url = (
+            response['hdurl'] if bool(params['hd']) is True
```
Comment (AlinaBlahun) [inline api/apod.py]: Please, don't compare boolean values to True or False using == or is operator. Read and use https://www.python.org/dev/peps/pep-0008/ it’s really easier for other developers to read your code when you do so.

File: tests/test_apod.py [hunk 17]
```diff
@@ -0,0 +1,53 @@
+from http import HTTPStatus
```
Comment (AlinaBlahun) [inline tests/test_apod.py:1]: A unit test should be independent, isolated, and repeatable. Read more about the unit testing idea, please (something like [this](https://medium.com/chris-nielsen/so-whats-a-good-unit-test-look-like-71f750333ac0#:~:text=Good%20unit%20tests%20are%20independent%20and%20isolated&text=They%20certainly%20don't%20rely,file%20system%20or%20a%20database.)) and try to fix yours.

File: api/errors.py [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+            'message': self.message,
+        }
+
+
+class AuthorizationError(ApiKeyError):
+    def __init__(self):
+        super().__init__('Invalid api key', HTTPStatus.BAD_REQUEST)
+
+
+class RequestError(ApiKeyError):
```
Comment (AlinaBlahun) [inline api/errors.py]: Strange naming, there is nothing about ApiKey at ApiKeyError. Maybe it's better to rename ApiKeyError to just ApiError or BaseError?

File: tests/test_apod.py (hunk 17 above)
Comment (mstoro) [reply tests/test_apod.py:1]: Hi, fixed tests. 
https://github.com/mstoro/nasa_api/pull/1/commits/6981799217867c0f50baa0dd1d387a94bf8aa070

File: tests/test_apod.py [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.json, expected_result)
+
+    @patch('requests.get')
+    def test_dayimage_invalid_apod_data(self, mock_get):
+        mock_get.return_value = self.get_mock_requests(
+            json=self.wrong_request_params,
+            status=HTTPStatus.OK
+        )
```
Comment (AlinaBlahun) [inline tests/test_apod.py]: It looks like the code doesn't reach that part with request in this case, so we don't really need mock here.

Comment (mstoro) [reply tests/test_apod.py]: https://github.com/mstoro/nasa_api/pull/1/commits/d778c8a10df1a0b33847b73b2c576ea208d5dc5b

**Review by AlinaBlahun — APPROVED** [review]
Good job!

**Pull request merged**